3.7-rc/alpine/docker-entrypoint.sh

```diff
@@ -390,6 +406,17 @@ if [ "$1" = 'rabbitmq-server' ] && [ "$shouldWriteConfig" ]; then
 			rabbit_set_config 'management.load_definitions' "$managementDefinitionsFile"
 		fi
 	fi
+
+	# if prometheus plugin is installed, generate config for it
```

Comment:
I am not sure how much of the <kbd>rabbitmq_prometheus</kbd> config should be generate here, but `prometheus.return_per_object_metrics = true` is especially useful.

FWIW, this is a more compact version of all supported config properties: https://github.com/rabbitmq/rabbitmq-prometheus#configuration

Reply:
If the goal is to remove this script and deprecate the logic in it, I think we should go the other direction and remove all `prometheus`-related support from this script/PR (which I'm working on now).

This means if folks want `rabbitmq_prometheus` to use TLS, they'll have to configure it to do so, but I think that's probably a good conversation starter for them to do the other TLS configuration themselves directly too (instead of using our helper environment variables) because that's the end goal anyhow.

Thoughts?

Reply:
More concretely, here's what I'm proposing this PR be simplified to (showing only the template changes):

<details>
<summary>Diff:</summary>

```diff
diff --git a/Dockerfile-alpine.template b/Dockerfile-alpine.template
index b255bd4..f29ac98 100644
--- a/Dockerfile-alpine.template
+++ b/Dockerfile-alpine.template
@@ -174,9 +174,9 @@ ENV RABBITMQ_DATA_DIR=/var/lib/rabbitmq
 RUN set -eux; \
 	addgroup -g 101 -S rabbitmq; \
 	adduser -u 100 -S -h "$RABBITMQ_DATA_DIR" -G rabbitmq rabbitmq; \
-	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
 	ln -sf "$RABBITMQ_DATA_DIR/.erlang.cookie" /root/.erlang.cookie
 
 # Use the latest stable RabbitMQ release (https://www.rabbitmq.com/download.html)
@@ -229,6 +229,8 @@ RUN set -eux; \
 	su-exec rabbitmq rabbitmq-plugins list; \
 # no stale cookies
 	rm "$RABBITMQ_DATA_DIR/.erlang.cookie"
+# Enable Prometheus-style metrics by default (https://github.com/docker-library/rabbitmq/issues/419)
+RUN set -eux; rabbitmq-plugins enable --offline rabbitmq_prometheus; echo 'management_agent.disable_metrics_collector = true' > /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # Added for backwards compatibility - users can simply COPY custom plugins to /plugins
 RUN ln -sf /opt/rabbitmq/plugins /plugins
@@ -246,5 +248,5 @@ ENV LANG=C.UTF-8 LANGUAGE=C.UTF-8 LC_ALL=C.UTF-8
 COPY docker-entrypoint.sh /usr/local/bin/
 ENTRYPOINT ["docker-entrypoint.sh"]
 
-EXPOSE 4369 5671 5672 25672
+EXPOSE 4369 5671 5672 15691 15692 25672
 CMD ["rabbitmq-server"]
diff --git a/Dockerfile-management.template b/Dockerfile-management.template
index 0177890..febf166 100644
--- a/Dockerfile-management.template
+++ b/Dockerfile-management.template
@@ -1,6 +1,8 @@
 FROM %%FROM%%
 
 RUN rabbitmq-plugins enable --offline rabbitmq_management
+# make sure the metrics collector is re-enabled (disabled in the base image for Prometheus-style metrics by default)
+RUN rm -f /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # extract "rabbitmqadmin" from inside the "rabbitmq_management-X.Y.Z.ez" plugin zipfile
 # see https://github.com/docker-library/rabbitmq/issues/207
diff --git a/Dockerfile-ubuntu.template b/Dockerfile-ubuntu.template
index e939ee7..84de22a 100644
--- a/Dockerfile-ubuntu.template
+++ b/Dockerfile-ubuntu.template
@@ -185,9 +185,9 @@ ENV RABBITMQ_DATA_DIR=/var/lib/rabbitmq
 RUN set -eux; \
 	groupadd --gid 999 --system rabbitmq; \
 	useradd --uid 999 --system --home-dir "$RABBITMQ_DATA_DIR" --gid rabbitmq rabbitmq; \
-	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
-	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	mkdir -p "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chown -fR rabbitmq:rabbitmq "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
+	chmod 777 "$RABBITMQ_DATA_DIR" /etc/rabbitmq /etc/rabbitmq/conf.d /tmp/rabbitmq-ssl /var/log/rabbitmq; \
 	ln -sf "$RABBITMQ_DATA_DIR/.erlang.cookie" /root/.erlang.cookie
 
 # Use the latest stable RabbitMQ release (https://www.rabbitmq.com/download.html)
@@ -247,6 +247,8 @@ RUN set -eux; \
 	gosu rabbitmq rabbitmq-plugins list; \
 # no stale cookies
 	rm "$RABBITMQ_DATA_DIR/.erlang.cookie"
+# Enable Prometheus-style metrics by default (https://github.com/docker-library/rabbitmq/issues/419)
+RUN set -eux; rabbitmq-plugins enable --offline rabbitmq_prometheus; echo 'management_agent.disable_metrics_collector = true' > /etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf
 
 # Added for backwards compatibility - users can simply COPY custom plugins to /plugins
 RUN ln -sf /opt/rabbitmq/plugins /plugins
@@ -264,5 +266,5 @@ ENV LANG=C.UTF-8 LANGUAGE=C.UTF-8 LC_ALL=C.UTF-8
 COPY docker-entrypoint.sh /usr/local/bin/
 ENTRYPOINT ["docker-entrypoint.sh"]
 
-EXPOSE 4369 5671 5672 25672
+EXPOSE 4369 5671 5672 15691 15692 25672
 CMD ["rabbitmq-server"]
diff --git a/update.sh b/update.sh
index c3d66e7..365485f 100755
--- a/update.sh
+++ b/update.sh
@@ -145,5 +145,17 @@ for version in "${versions[@]}"; do
 			-e "s!%%INSTALL_PYTHON%%!$installPython!g" \
 			Dockerfile-management.template \
 			> "$version/$variant/management/Dockerfile"
+
+		# rabbitmq_prometheus is only included in 3.8+
+		# same with "/etc/rabbitmq/conf.d" support (https://github.com/rabbitmq/rabbitmq-server/pull/2258, https://github.com/rabbitmq/rabbitmq-server/pull/2277)
+		if [ "$rcVersion" = '3.7' ]; then
+			sed -ri \
+				-e '/[Pp]rometheus/d' \
+				-e '/metrics_collector/d' \
+				-e 's/ 15691 15692 / /g' \
+				-e 's! /etc/rabbitmq/conf.d ! !g' \
+				"$version/$variant/Dockerfile" \
+				"$version/$variant/management/Dockerfile"
+		fi
 	done
 done
```

</details>

Separately, I'd like to also make a PR to start showing a large warning to users for whom we're currently generating a configuration file so that they have some way to know that we intend to remove most of this script's functionality at some later date.

Reply:
> If the goal is to remove this script and deprecate the logic in it, I think we should go the other direction and remove all prometheus-related support from this script/PR (which I'm working on now).

Yes, agreed.

> More concretely, here's what I'm proposing this PR be simplified to

Looking good!

The only issue that I can foresee is if someone uses the non-management image and enables the management plugin. In that case, they will need to manually delete `/etc/rabbitmq/conf.d/management_agent.disable_metrics_collector.conf` in order for the UI to look as expected - it's super fast and bare-bones without all the metrics and graphs. Not a biggie and mostly unlikely, but worth a mention.

> deprecation strawman PR is up at #424

I will chip in, thanks for kicking it off 👍